go/vt/vttablet/tabletmanager/vreplication/vcopier.go

```diff
@@ -364,3 +640,522 @@ func (vc *vcopier) fastForward(ctx context.Context, copyState map[string]*sqltyp
+func (vrh *vcopierCopyTaskResultHooks) notify(ctx context.Context, result *vcopierCopyTaskResult) {
+	for _, fn := range vrh.fns {
+		fn(ctx, result)
+	}
+}
+
+// sendTo registers a hook that accepts a result and sends the result to the
+// provided channel. E.g.:
+//
+//	resultCh := make(chan *vcopierCopyTaskResult, 1)
+//	task.lifecycle.onResult().sendTo(resultCh)
+//	defer func() {
+//	  result := <-resultCh
+//	}()
+func (vrh *vcopierCopyTaskResultHooks) sendTo(ch chan<- *vcopierCopyTaskResult) {
+	vrh.do(func(ctx context.Context, result *vcopierCopyTaskResult) {
+		select {
+		case ch <- result:
+		case <-ctx.Done():
+			// Failure to send the result to the consumer on the other side of
+			// the channel before context expires will have the following
+			// consequences:
+			//
+			// * Subsequent tasks waiting for this task to complete won't run.
+			//   That's OK. They won't hang waiting on the channel because,
+			//   like this task they respond to context expiration.
+			// * The outermost loop managing task execution may not know that
+			//   this task failed or succeeded.
+			//   - In the case that this task succeeded, statistics and logging
+			//     will not indicate that this task completed. That's not great,
+			//     but shouldn't negatively impact the integrity of data or the
+			//     copy workflow.
+			//   - In the case that this task failed, there should be no adverse
+			//     impact: the outermost loop handles context expiration by
+			//     stopping the copy phase without completing it.
+		}
+	})
+}
+
+// awaitCompletion registers a callback that returns an error unless the
+// provided chan produces a vcopierTaskResult in a complete state.
+//
+// This is useful for sequencing vcopierCopyTasks, e.g.:
+//
+//	resultCh := make(chan *vcopierCopyTaskResult, 1)
+//	prevT.lifecycle.onResult().sendTo(resultCh)
+//	currT.Lifecycle.before(vcopierCopyTaskUpdateCopyState).awaitCompletion(resultCh)
+func (vth *vcopierCopyTaskHooks) awaitCompletion(resultCh <-chan *vcopierCopyTaskResult) {
+	vth.do(func(ctx context.Context, args *vcopierCopyTaskArgs) error {
+		select {
+		case result := <-resultCh:
+			if result == nil {
+				return fmt.Errorf("channel was closed before a result received")
+			}
+			if !vcopierCopyTaskStateIsDone(result.state) {
+				return fmt.Errorf("received result is not done")
+			}
+			if result.state != vcopierCopyTaskComplete {
+				return fmt.Errorf("received result is not complete")
+			}
+			return nil
+		case <-ctx.Done():
+			// A context expiration probably indicates a PlannedReparentShard
+			// or an elapsed copy phase duration. Those aren't treated as error
+			// conditions, but we'll return the context error here anyway.
+			//
+			// Task execution will detect the presence of the error, mark this
+			// task canceled, and abort. Subsequent tasks won't execute because
+			// this task didn't complete.
+			return ctx.Err()
```

Comment:
We should follow [the established convention](https://github.com/vitessio/vitess/search?q=Code_Canceled) of returning a `vterror` with the Canceled code:
```
return vterrors.Errorf(vtrpcpb.Code_CANCELED, "context has expired")
```